cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -1282,6 +1293,254 @@ def _check_column_availability(self):
+        checks are executed separately) but no additional keys are allowed.
+
+        :param dict dictionary: the dictionary to check.
+        :param List[str] allowed_keys: the list of allowed keys.
+
+        :rtype bool
+        """
+        return set(dictionary.keys()).issubset(allowed_keys)
+
+    def _is_supported_spatial_assay(self):
+        """
+        Determine if the assay_ontology_term_id is either Visium (EFO:0010961) or Slide-seqV2 (EFO:0030062).
+
+        :return True if assay_ontology_term_id is Visium or Slide-seqV2, False otherwise.
+        :rtype bool
+        """
+        return self.adata.obs.get("assay_ontology_term_id").isin([ASSAY_VISIUM, ASSAY_SLIDE_SEQV2]).any()
+
+    def _is_valid_visium_image_shape(self, image: np.ndarray) -> bool:
+        """
+        Determine if the image has shape (,,3); image is expected to be a 3D numpy array
+        with the size of the last dimension being three.
+
+        :param np.ndarray image: the image to check the shape of.
+
+        :return True if image has shape (,,3), False otherwise.
+        :rtype bool
+        """
+        return len(image.shape) == 3 and image.shape[2] == 3
+
+    def _is_visium(self) -> bool:
+        """
+        Determine if the assay_ontology_term_id is Visium (EFO:0010961).
+
+        :return True if assay_ontology_term_id is Visium, False otherwise.
+        :rtype bool
+        """
+        assay_ontology_term_id = self.adata.obs.get("assay_ontology_term_id")
+        return assay_ontology_term_id is not None and (assay_ontology_term_id == ASSAY_VISIUM).any()
+
+    def _validate_spatial_image_shape(self, image_name: str, image: np.ndarray, max_dimension: int = None):
+        """
+        Validate the spatial image is of shape (,,3) and has a max dimension, if specified. A spatial image
+        is either spatial[library_id]['images']['hires'] or spatial[library_id]['images']['fullres']. Errors
+        are added to self.errors if any.
+
+        :param str image_name: the name of the image, either "hires" or "fullres".
+        :param np.ndarray image: the image to validate.
+        :param int max_dimension: the maximum dimension of the image, optional.
+
+        :rtype None
+        """
+        # Image must be an ndarray.
+        if not isinstance(image, np.ndarray):
+            self.errors.append(
+                f"uns['spatial'][library_id]['images']['{image_name}'] must be of numpy.ndarray type, "
+                f"it is {type(image)}."
+            )
+            return
+
+        # Confirm shape of image is valid: allowed shape is (,,3).
+        if not self._is_valid_visium_image_shape(image):
+            self.errors.append(
+                f"uns['spatial'][library_id]['images']['{image_name}'] must have shape (,,3), "
+                f"it has shape {image.shape}."
+            )
+
+        # Confirm max dimension of image, if specified, is valid.
+        if max_dimension is not None and max(image.shape) > max_dimension:
```

Comment:
https://github.com/chanzuckerberg/single-cell-curation/pull/858#issuecomment-2078156945

looks like we need to update this check + the accompanying comments to reflect that we want the max dimension to == max_dimension exactly